File: common/src/main/java/io/xol/chunkstories/net/PacketsStore.java
//
// This file is a part of the Chunk Stories Implementation codebase
// Check out README.md for more information
// Website: http://chunkstories.xyz
//

package io.xol.chunkstories.net;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.util.HashMap;
import java.util.Iterator;
import java.util.Map;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import io.xol.chunkstories.api.content.Asset;
import io.xol.chunkstories.api.content.Content;
import io.xol.chunkstories.api.exceptions.content.IllegalPacketDeclarationException;
import io.xol.chunkstories.api.exceptions.net.UnknowPacketException;
import io.xol.chunkstories.api.net.Packet;
import io.xol.chunkstories.api.net.PacketDefinition;
import io.xol.chunkstories.content.GameContentStore;

public class PacketsStore implements Content.PacketDefinitions {

	private final GameContentStore store;
	private final Map<String, PacketDefinitionImplementation> byNames = new HashMap<String, PacketDefinitionImplementation>();
	private final Map<Class<? extends Packet>, PacketDefinitionImplementation> byClasses = new HashMap<Class<? extends Packet>, PacketDefinitionImplementation>();

	// private PacketDefinition textPacket, filePacket;

	private static final Logger logger = LoggerFactory.getLogger("content.packets");

	public Logger logger() {
		return logger;
	}

	public PacketsStore(GameContentStore store) {
		this.store = store;

		// reload();
	}

	public void reload() {

		byNames.clear();
		byClasses.clear();

		// Load system.packets
		InputStream is = getClass().getResourceAsStream("/system.packets");
		readPacketsDefinitions(new BufferedReader(new InputStreamReader(is)), new Object() {
			@Override
			public String toString() {
				return "system.packets";
			}
		});
		//TODO review this system
	}

	private void readPacketsDefinitions(Asset f) {
		if (f == null)
			return;
		BufferedReader reader = new BufferedReader(f.reader());
		readPacketsDefinitions(reader, f);
	}

	private void readPacketsDefinitions(BufferedReader reader, Object source) {
		logger().debug("Reading packets definitions in : " + source);

		//TODO
		//throw new UnsupportedOperationException("TODO");
		/*try {
			String line = "";
			PacketDefinitionImplementation packetType = null;
			while ((line = reader.readLine()) != null) {
				line = line.replace("\t", "");
				if (line.startsWith("#")) {
					// It's a comment, ignore.
				}
				// We shouldn't come accross end tags by ourselves, this is
				// dealt with in the constructors
				else if (line.startsWith("end")) {
					logger().warn("Syntax error in file : " + source + " : Unexpected 'end' tag.");
					continue;
				} else if (line.startsWith("packet")) {
					if (line.contains(" ")) {
						String[] split = line.split(" ");
						String materialName = split[1];

						try {
							packetType = new PacketDefinitionImplementation(store, materialName, reader);
						} catch (IllegalPacketDeclarationException e) {
							store.logger().error(e.getMessage());
							continue;
						}

						// Eventually add the packet type
						byNames.put(packetType.getName(), packetType);

						// Add quick-resolve hashmap entries
						if (packetType.clientClass != null)
							this.byClasses.put(packetType.clientClass, packetType);
						if (packetType.serverClass != null)
							this.byClasses.put(packetType.serverClass, packetType);
						if (packetType.commonClass != null)
							this.byClasses.put(packetType.commonClass, packetType);
					}
				}
			}
			reader.close();
		} catch (IOException e) {
			e.printStackTrace();
		}*/
	}

	@Override
	public PacketDefinition getPacketByName(String name) {
		return byNames.get(name);
	}

	@Override
	public PacketDefinition getPacketFromInstance(Packet packet) throws UnknowPacketException {
		Class<? extends Packet> pclass = packet.getClass();

		PacketDefinitionImplementation ptd = this.byClasses.get(pclass);
		if (ptd != null)
			return ptd;

		throw new UnknowPacketException(packet);
	}

	@Override
	public Content parent() {
		return this.store;
	}

	@Override
	public Iterator<PacketDefinition> all() {
		Iterator<PacketDefinitionImplementation> i = byNames.values().iterator();
		return new Iterator<PacketDefinition>() {

			@Override
			public boolean hasNext() {
				return i.hasNext();
			}

			@Override
			public PacketDefinition next() {
				return i.next();
			}

		};
	}

	/*
	 * @Override public PacketDefinition getTextPacket() { return textPacket; }
	 * 
	 * @Override public PacketDefinition getFilePacket() { return filePacket; }
	 */
}

File: common/src/main/java/xyz/chunkstories/net/PacketsStore.kt
//
// This file is a part of the Chunk Stories Implementation codebase
// Check out README.md for more information
// Website: http://chunkstories.xyz
//

package xyz.chunkstories.net

import com.google.gson.Gson
import org.hjson.JsonValue
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import xyz.chunkstories.api.content.json.asDict
import xyz.chunkstories.api.net.Packet
import xyz.chunkstories.content.GameContentStore
import xyz.chunkstories.content.eat
import java.io.Reader
import java.util.*

class PacketsStore(val parent: GameContentStore) {
    private val byNames = HashMap<String, PacketDefinition>()
    private val byClasses = HashMap<Class<out Packet>, PacketDefinition>()

    fun reload() {
        byNames.clear()
        byClasses.clear()

        val gson = Gson()

        fun readDefinitions(r: Reader) {
            val json = JsonValue.readHjson(r).eat().asDict ?: throw Exception("This json isn't a dict")
            val dict = json["packets"].asDict ?: throw Exception("This json doesn't contain an 'packets' dict")

            for (element in dict.elements) {
                val name = element.key
                val properties = element.value.asDict ?: throw Exception("Definitions have to be dicts")

                val packetDefinition = PacketDefinition(parent, name, properties)
                byNames[name] = packetDefinition

                packetDefinition.commonClass?.let { byClasses[it] = packetDefinition }
                packetDefinition.serverClass?.let { byClasses[it] = packetDefinition }
                packetDefinition.clientClass?.let { byClasses[it] = packetDefinition }

                logger.debug("Loaded packet definition $packetDefinition")
            }
        }

        // Load system.packets
        readDefinitions(javaClass.getResourceAsStream("/packets/systemPackets.hjson").reader())

        // Load the rest
        for (asset in parent.modsManager.allAssets.filter { it.name.startsWith("packets/") && it.name.endsWith(".hjson") }) {
            logger.debug("Reading packets definitions in : $asset")
            readDefinitions(asset.reader())
        }
    }

    fun getPacketByName(name: String): PacketDefinition? {
        return byNames[name]
    }

    fun getPacketFromInstance(packet: Packet): PacketDefinition? {
        val pclass = packet.javaClass

        val ptd = this.byClasses[pclass]
        if (ptd != null)
            return ptd

        return null
    }

    val all: Collection<PacketDefinition>
        get() {
            return byNames.values
        }

    companion object {
        private val logger = LoggerFactory.getLogger("content.packets")
    }

    val logger: Logger
        get() = Companion.logger
}
